lib/ipam/ipam.go

```diff
@@ -1458,27 +1472,50 @@ func (c ipamClient) IPsByHandle(ctx context.Context, handleID string) ([]net.IP,
 	return assignments, nil
 }
 
-// ReleaseByHandle releases all IP addresses that have been assigned
-// using the provided handle.
-func (c ipamClient) ReleaseByHandle(ctx context.Context, handleID string) error {
+// ReleaseByHandle releases addresses with the given handle using the given revision information.
+func (c ipamClient) ReleaseByHandle(ctx context.Context, handleID, revision string, uid *types.UID) error {
+	// Get the existing handle and populate the given revision and UID information.
 	handleID = sanitizeHandle(handleID)
 	log.Debugf("Releasing all IPs with handle '%s'", handleID)
-	obj, err := c.blockReaderWriter.queryHandle(ctx, handleID, "")
+	obj, err := c.blockReaderWriter.queryHandle(ctx, handleID, revision)
 	if err != nil {
 		return err
 	}
-	handle := allocationHandle{obj.Value.(*model.IPAMHandle)}
+	if revision != "" {
+		obj.Revision = revision
+	}
+	if uid != nil {
+		obj.UID = uid
+	}
 
-	for blockStr := range handle.Block {
+	log.Infof("Releasing all IPs with handle '%s', RV=%s; UID=%v", handleID, obj.Revision, obj.UID)
+	h := allocationHandle{obj.Value.(*model.IPAMHandle)}
+	for blockStr := range h.Block {
 		_, blockCIDR, _ := net.ParseCIDR(blockStr)
-		if err := c.releaseByHandle(ctx, handleID, *blockCIDR); err != nil {
+		if obj, err = c.releaseByHandle(ctx, obj, *blockCIDR); err != nil {
 			return err
 		}
 	}
+
+	if obj != nil {
+		// Delete the handle itself. This normally would have happened as a side-effect of
+		// decrementing the handle. However, if for some reason there are no IPs allocated with this handle,
+		// we will never decrement it and thus it will never otherwise be deleted.
+		if err = c.blockReaderWriter.deleteHandle(ctx, obj); err != nil {
+			if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+				// We expect the handle to either not exist, or be deleted cleanly.
+				// If it's not either of these things, return an error.
+				return err
+			}
+		}
+	}
 	return nil
 }
 
-func (c ipamClient) releaseByHandle(ctx context.Context, handleID string, blockCIDR net.IPNet) error {
+// releaseByHandle releases all addresses with the given handle object, and returns an updated handle object or nil if
+// the handle was deleted, or nil and an error if an error was encountered.
+func (c ipamClient) releaseByHandle(ctx context.Context, handle *model.KVPair, blockCIDR net.IPNet) (*model.KVPair, error) {
+	handleID := handle.Key.(model.IPAMHandleKey).HandleID
```

Comment:
I think this retry processing is not valid anymore if passing in revision and UUID.

Who would now handle this retry?  The calling code?   Wondering if passing in the revision is correct.
Is the IPAM data going to contain a UUID and handle?